ipywidgets/widgets/widget_dragdrop.py

```diff
@@ -0,0 +1,82 @@
+# Copyright (c) Jupyter Development Team.
+# Distributed under the terms of the Modified BSD License.
+
+"""Contains the DropWidget class"""
+from .widget import Widget, CallbackDispatcher, register, widget_serialization
+from .domwidget import DOMWidget
+from .widget_core import CoreWidget
+from traitlets import Bool, Dict, Unicode, Instance
+
+
+class DropWidget(DOMWidget, CoreWidget):
+    """Widget that has the ondrop handler. Used as a mixin"""
+
+    draggable = Bool(default=False).tag(sync=True)
+    drag_data = Dict().tag(sync=True)
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self._drop_handlers = CallbackDispatcher()
+        self._dragstart_handlers = CallbackDispatcher()
+        self.on_msg(self._handle_dragdrop_msg)
+
+    def on_drop(self, callback, remove=False):
+        """
+        Register a callback to execute when an element is dropped.
+        The callback will be called with two arguments, the clicked button
+        widget instance, and the dropped element data.
+        Parameters
+        ----------
+        remove: bool (optional)
+            Set to true to remove the callback from the list of callbacks.
+        """
+        self._drop_handlers.register_callback(callback, remove=remove)
+
+    def drop(self, data):
+        """
+        Programmatically trigger a drop event.
+        This will call the callbacks registered to the  drop event.
+        """
+
+        if data.get('application/vnd.jupyter.widget-view+json'):
```

Comment:
Small coding style comment:

Maybe `application/vnd.jupyter.widget-view+json` should be in an all-caps variable that we reuse everywhere the mime type is used?

Reply:
I am actually not following the spec here correctly (which should be a small dictionary with `{model: XYZ, major: 0, minor: 2}` -- the reason is that setData only allows for strings. Obviously we could encode the JSON as a string, and decode it on the python side but I didn't think it would be worth it...

Reply:
cc @jasongrout you may have an opinion on whether we should write the JSON into a string on merely pass the id.

Reply:
(regardless if we use that string constant in multiple place it should probably be in a variable.)

Reply:
> cc @jasongrout you may have an opinion on whether we should write the JSON into a string on merely pass the id.

I'm a bit confused by what this data is, but certainly a model id is much simpler. I strongly feel that if we use the mimetype, we should be consistent and use it exactly as we have specified.

Reply:
Thedrag and drop API specifies a mime type key but the value is a string.

Reply:
Ok I switched to the real JSON type now, and do de-serialization from JSON in the backend.